ext/node/path/posix.rs

```diff
@@ -0,0 +1,47 @@
+// Copyright 2018-2023 the Deno authors. All rights reserved. MIT license.
+
+use deno_core::error::AnyError;
+use deno_core::op;
+use deno_core::serde_json::Value;
+
+use super::assert_path;
+use super::char_at_index;
+use super::constants;
+
+#[op]
+pub fn op_node_path_posix_dirname(path: Value) -> Result<String, AnyError> {
+  let path = assert_path(&path)?;
+  if path.chars().count() == 0 {
+    return Ok(String::from("."));
+  }
+
+  let has_root =
+    char_at_index(path, 0).unwrap() as u32 == constants::CHAR_FORWARD_SLASH;
+  let mut end: Option<usize> = None;
+  let mut matched_slash = true;
+
+  for (i, c) in path.char_indices().rev() {
+    if i < 1 {
+      break;
+    }
+    match c as u32 {
+      constants::CHAR_FORWARD_SLASH if !matched_slash => {
+        end = Some(i);
+        break;
+      }
+      constants::CHAR_FORWARD_SLASH => {
+        // noop
+      }
+      _ => matched_slash = false,
+    }
+  }
+
+  let dirname = match (end, has_root) {
+    (None, true) => String::from("/"),
+    (None, false) => String::from("."),
+    (Some(1), true) => String::from("//"),
+    (Some(ending), _) => path.chars().take(ending).collect(),
```

Comment:
path[0..ending].to_string()

Reply:
getting a substring this way returns 0 to N bytes of the path string (where n = the `ending` var in this example). This is an issue because if the string contains multi-byte characters than you'll get broken characters. If you have a file path with an emoji for example, that would break.